ext/mysql2/result.c

```diff
@@ -920,6 +920,9 @@ static VALUE rb_mysql_result_each(int argc, VALUE * argv, VALUE self) {
     wrapper->numberOfRows = wrapper->stmt_wrapper ? mysql_stmt_num_rows(wrapper->stmt_wrapper->stmt) : mysql_num_rows(wrapper->result);
     wrapper->rows = rb_ary_new2(wrapper->numberOfRows);
   } else if (wrapper->rows && !cacheRows) {
+    if (wrapper->resultFreed) {
+      rb_raise(cMysql2Error, "Statement result already freed");
+    }
```

Comment:
Nice find. I would raise this higher up in the function - just after the check for `"Statement handle already closed"` above.

That said, I wonder what would cause this. Are you using the word "Statement" here specifically because this only occurs for prepared statements, or would the word "Query" be more appropriate?